batavia/types/StrUtils.js

```diff
@@ -3,9 +3,53 @@ var type_name = require('../core').type_name;
 var BigNumber = require('bignumber.js').BigNumber
 
 function _substitute(format, args){
-
     var types = require('../types');
-    var workingArgs = args.slice();
+
+    // INITIAL SETUP
+    // does this conversion use key word or sequential args?
+    var kwRe = /%\(/;
+    var usesKwargs = (format.match(kwRe) !== null)
+
+    // if using kwargs, fail if first arg isn't a dict or more than 1 arg given
+    if ( usesKwargs && (
+            !types.isinstance(args[0], types.Dict) || args.length != 1
+        )
+    ){
+        throw new exceptions.TypeError("format requires a mapping")
+    }
+
+    function Args(args, usesKwargs) {
+        // represents either sequential or key word arguments.
+        // functions that need to grab arguments interface with this object
+        if (usesKwargs){
+            this.collection = args[0]
+        } else{
+            this.remainingArgs = args.slice(); // remaining args
+        }
+
+        this.getArg = function(key){
```

Comment:
Ideally this would be added to the prototype, rather than an attribute on the instance.